File: 287-find-the-duplicate-number/find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int slow = 0, fast = 0;
        while(slow != fast || (slow == 0 && fast == 0)) {
            fast = nums[nums[fast]];
            slow = nums[slow];
        }
        slow = 0;
        while(slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
}

File: 287-find-the-duplicate-number/find-the-duplicate-number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function(nums) {
    let fast = 0;
    let slow = 0;
    while(true) {
        fast = nums[nums[fast]];
        slow = nums[slow];
        if(nums[slow] === nums[fast]) break;
    }
    let first = 0;
    while(first !== slow) {
        first = nums[first];
        slow = nums[slow];
    }
    return slow;
};